lib/veres-one-did-doc.js

```diff
@@ -0,0 +1,444 @@
+/*!
+ * Copyright (c) 2018 Veres One Project. All rights reserved.
+ */
+'use strict';
+
+const constants = require('./constants');
+const {LDKeyPair} = require('crypto-ld');
+
+const uuid = require('uuid/v4');
+const jsonpatch = require('fast-json-patch');
+
+const VERES_DID_REGEX = /^(did\:v1\:)(test\:)?(uuid|nym)\:(.+)/;
+
+class VeresOneDidDoc {
+  constructor(options = {}) {
+    this.injector = options.injector;
+
+    this.doc = options.doc || {'@context': constants.VERES_ONE_V1_CONTEXT};
+    this.meta = options.meta || {sequence: 0};
+
+    this.observer = null; // JSON Patch change observer
+
+    // Includes private keys -- this property will not be serialized.
+    this.keys = options.keys || {};
+  }
+
+  /**
+   * Generates a new DID Document and initializes various authentication
+   * and authorization suite keys.
+   *
+   * @param options
+   * @param [options.injector]
+   *
+   * Params needed for DID generation:
+   * @param [options.didType='nym'] {string} DID type, 'nym' or 'uuid'
+   * @param [options.env] {string} 'dev'/'live' etc.
+   *
+   * Params needed for key generation:
+   * @param [options.keyType] {string}
+   * @param [options.passphrase] {string}
+   *
+   * @throws {Error}
+   *
+   * @returns {VeresOneDidDoc}
+   */
+  static async generate(options) {
+    const keyType = options.keyType || constants.DEFAULT_KEY_TYPE;
+    if(!constants.SUPPORTED_KEY_TYPES.includes(keyType)) {
+      throw new Error(`Unknown key type: "${keyType}"`);
+    }
+
+    const did = new VeresOneDidDoc(options);
+    await did.init({keyType, ...options});
+    return did;
+  }
+
+  /**
+   * Returns the DID uri.
+   */
+  get id() {
+    return this.doc.id;
+  }
+
+  set id(value) {
```

Comment:
The reason it's a setter (and a corresponding getter) is because the id is delegated to the `did.doc` object. That said, the only time I'm modifying the ID is in unit tests, so I can remove the setter.

Reply:
Setter removed.